sandbox/cgroup/freezer.go

```diff
@@ -100,6 +113,100 @@ func thawSnapProcessesImpl(snapName string) error {
 	return nil
 }
 
+func applyToSnap(snapName string, action func(groupName string) error, skipErrs bool) error {
+	canary := fmt.Sprintf("snap.%s.", snapName)
+	cgroupRoot := filepath.Join(rootPath, cgroupMountPoint)
+	if _, dir, _ := osutil.DirExists(cgroupRoot); !dir {
+		return nil
+	}
+	return filepath.Walk(filepath.Join(rootPath, cgroupMountPoint), func(name string, info os.FileInfo, err error) error {
+		if err != nil && !skipErrs {
+			return err
+		}
+		if !info.IsDir() {
+			return nil
+		}
+		if !strings.HasPrefix(info.Name(), canary) {
+			return nil
+		}
+		// found a group
+		if err := action(name); err != nil && !skipErrs {
+			return err
+		}
+		return filepath.SkipDir
+	})
+}
+
+// freezeSnapProcessesImplV2 freezes all the processes originating from the
+// given snap. Processes are frozen regardless of which particular snap
+// application they originate from.
+func freezeSnapProcessesImplV2(snapName string) error {
+	// in case of v2, the process calling this code, (eg. snap-update-ns)
+	// may already be part of the trackign cgroup for particular snap, care
+	// must be taken to not freeze ourselves
+	ownGroup, err := cgroupProcessPathInTrackingCgroup(os.Getpid())
+	if err != nil {
+		return err
+	}
+	ownGroupDir := filepath.Join(rootPath, cgroupMountPoint, ownGroup)
+	freezeOne := func(dir string) error {
+		if dir == ownGroupDir {
+			// let's not freeze ourselves
+			logger.Debugf("freeze, skipping own group %v", dir)
+			return nil
+		}
+		fname := filepath.Join(dir, "cgroup.freeze")
+		if err := ioutil.WriteFile(fname, []byte("1"), 0644); err != nil && os.IsNotExist(err) {
+			//  the group may be gone already
+			return nil
+		} else if err != nil {
+			return fmt.Errorf("cannot freeze processes of snap %q, %v", snapName, err)
+		}
+		for i := 0; i < 30; i++ {
+			data, err := ioutil.ReadFile(fname)
+			if err != nil {
+				return fmt.Errorf("cannot determine the freeze state of processes of snap %q, %v", snapName, err)
+			}
+			// If the cgroup is still freezing then wait a moment and try again.
+			if bytes.Equal(bytes.TrimSpace(data), []byte("1")) {
+				// we're done
+				return nil
+			}
+			// add a bit of delay
+			time.Sleep(100 * time.Millisecond)
+		}
+		return fmt.Errorf("cannot freeze processes of snap %q in group %v", snapName, filepath.Base(dir))
+	}
+	const skipFreezeErrs = false
+	err = applyToSnap(snapName, freezeOne, skipFreezeErrs)
+	if err == nil {
+		return nil
+	}
+	// thaw one by one?
```

Comment:
It might be useful to have a similar comment here as in the v1 code, something like:

	// If we got here then we timed out after waiting for freezing for too long.